dcos/recordio.py

```diff
@@ -0,0 +1,164 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Provides facilities for "Record-IO" encoding of data.
+"Record-IO" encoding allows one to encode a sequence
+of variable-length records by prefixing each record
+with its size in bytes:
+
+5\n
+hello
+6\n
+world!
+
+Note that this currently only supports record lengths
+encoded as base 10 integer values with newlines as a
+delimiter. This is to provide better language portability
+portability: parsing a base 10 integer is simple. Most
+other "Record-IO" implementations use a fixed-size header
+of 4 bytes to directly encode an unsigned 32 bit length.
+"""
+
+from dcos.errors import DCOSException
+
+
+class Encoder(object):
+    """Encode an arbitray message type into a 'RecordIO' message.
+
+       This class encapsulates the process of encoding an
+       arbitrary string of bytes into a 'RecordIO' message. Its
+       constructor takes a serialization function of the form
+       'serialize(message)'. This serialization function is
+       responsible for knowing how to take whatever message type
+       is passed to 'encode()' and serializing it to a string.
+
+       Once 'encode(message)' is called, it will use the
+       serialization function to convert 'message' into a string
+       of bytes, wrap that string fo bytes in a 'RecordIO' frame,
+       and return it.
+    """
+
+    def __init__(self, serialize):
+        """Initializes a 'RecordIO' encoder'.
+
+        ":param serialize: a function to serialize any message
+                           passed to 'encode()'
+        type serialize: function
+        """
+
+        self.serialize = serialize
+
+    def encode(self, message):
+        """Encode a message into 'RecordIO' format.
+
+        ":param message: a message to serialize and then wrap in
+                       a 'RecordIO' frame.
+        type message: object
+        returns: a serialzied message wrapped in a 'RecordIO' frame.
+        rtype: str
+        """
+
+        s = self.serialize(message)
+        return str(len(s)) + "\n" + s
```

Comment:
I would just require the caller to send a `bytearray` so they can determine how the record is encoded.